### PR TITLE
Convert to use AsyncHttpClient

### PR DESCRIPTION
**Pull request by stevenschlansker**

### CODE REVIEW

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 1]
```diff
@@ -13,209 +13,228 @@
...
+    public EtcdClient(
             ObjectMapper mapper,
             URI baseUri)
     {
-        if (!client.getConfiguration().isRegistered(EtcdResponseInterceptor.class)) {
-            client.register(EtcdResponseInterceptor.class);
-        }
-        this.client = client;
+        this.client = new DefaultAsyncHttpClient(new DefaultAsyncHttpClientConfig.Builder()
+                .setFollowRedirect(true)
```
Comment (benhardy) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: wanna make these configurable?

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 2]
```diff
@@ -13,209 +13,228 @@
...
+                .setMethod(PUT)
+                .addFormParam("refresh", "true")
+                .addFormParam("prevExist", "true")
+                .addFormParam("ttl", Integer.toString(ttl)))
+            .toCompletableFuture()
+            .thenApply(this::handleResponse);
     }
 
-    public EtcdResult cas(String key, String expect, String set)
+    public CompletableFuture<EtcdResult> cas(String key, String expect, String set)
```
Comment (benhardy) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: what's this supposed to do?

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 3]
```diff
@@ -13,209 +13,228 @@
...
+        }
+
+        EtcdResult result;
+        try {
+            result = readResult(r);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+
+        if (code >= 400) {
```
Comment (benhardy) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: are non-404s likely here?

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 2 above)
Comment (stevenschlansker) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Compare and set, but we don't use it yet.  I guess I can just nuke it.

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 1 above)
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Should we new up the HTTP client or inject it as before?

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 4]
```diff
@@ -13,209 +13,228 @@
...
-        if (!client.getConfiguration().isRegistered(EtcdResponseInterceptor.class)) {
-            client.register(EtcdResponseInterceptor.class);
-        }
-        this.client = client;
+        this.client = new DefaultAsyncHttpClient(new DefaultAsyncHttpClientConfig.Builder()
+                .setFollowRedirect(true)
+                .setConnectTimeout(2000)
+                .setPooledConnectionIdleTimeout(30000)
+                .setMaxConnections(100)
+                .setMaxConnectionsPerHost(100)
```
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: What about a socket read timeout?

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 5]
```diff
@@ -13,209 +13,228 @@
...
 
-    private Builder getKeyInvocation(String key)
-    {
-        return getKeyTarget(key).request(MediaType.APPLICATION_JSON_TYPE);
-    }
+    private EtcdResult handleResponse(Response r) {
+        int code = r.getStatusCode();
+        if (code == 404) {
+            return null;
+        }
```
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Why was the previous 404 handling not preferable?

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 2 above)
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: @benhardy Compare-and-set.

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 3 above)
Comment (stevenschlansker) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Not really.  If there's a 4xx error (excluding 404, which is handled on line 76), then the "real" problem is nested in the json response as the `EtcdResult.errorCode`

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 6]
```diff
@@ -13,209 +13,228 @@
...
+                    .get();
+            if (r.getStatusCode() != 200) {
+                throw new IOException("bad response: " + r.getStatusCode() + " " + r.getResponseBody());
+            }
+            return r.getResponseBody();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw Throwables.propagate(e);
+        } catch (ExecutionException e) {
+            throw Throwables.propagate(e.getCause());
```
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Still synchronous?

File: src/main/java/com/opentable/etcd/EtcdClient.java [hunk 7]
```diff
@@ -13,209 +13,228 @@
...
 
-    private ServerErrorException describeAndRethrow(ServerErrorException e) {
-        try {
-            final Response r = e.getResponse();
-            LOGGER.error("Server response {}: \"{}\"", r.getStatus(), r.readEntity(String.class));
-        } catch (Exception e1) {
-            LOGGER.error(String.format("While trying to describe exception %s", e), e1);
+    private static long parseLongHeader(Response r, String header)
+    {
+        return parseLongHeader(r, header, Long.MIN_VALUE);
```
Comment (pennello) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Is this default value always correct?

File: pom.xml [hunk 8]
```diff
@@ -29,7 +29,7 @@
 
   <groupId>com.opentable.components</groupId>
   <artifactId>otj-etcd-client</artifactId>
-  <version>0.13-SNAPSHOT</version>
+  <version>0.15-SNAPSHOT</version>
```
Comment (pennello) [inline pom.xml]: No semver?

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 4 above)
Comment (stevenschlansker) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Added.

File: src/main/java/com/opentable/etcd/EtcdClient.java (hunk 6 above)
Comment (stevenschlansker) [inline src/main/java/com/opentable/etcd/EtcdClient.java]: Yeah, nobody cares if `getVersion` is async or not.

File: pom.xml (hunk 8 above)
Comment (stevenschlansker) [inline pom.xml]: Not yet.  `1.0` will be the first semver release.  This is still in "I'm hacking on the API and it's not ready yet" mode.

**Comment by stevenschlansker** [comment]
> What's the general motivation for making the client async? Isn't etcd an internal service that we maintain?

Yes

> Are requests to services that use this client not dependent on the result being completed before they can return?

Some are, some aren't.  The observation was that during stress testing, etcd will group changes and write them in batches.  And then you have to wait for a disk flush.  So you may end up blocking for ~100ms waiting for your coincident writers to all get flushed.  This causes excessive thread usage in discovery; we'd rather be off servicing some other request than waiting for etcd to write.

**Comment by pennello** [comment]
Steven and I discussed in person.  Looks good, pending the following two changes:
- Make the max async connections tunable.
- Add metric for in-use connections.

**Pull request merged**